rustls/src/msgs/message.rs

```diff
@@ -406,65 +421,121 @@ impl TryFrom<PlainMessage> for Message<'static> {
 ///
 /// A [`PlainMessage`] must contain plaintext content. Encrypted content should be stored in an
 /// [`OpaqueMessage`] and decrypted before being stored into a [`PlainMessage`].
-impl<'a> TryFrom<BorrowedPlainMessage<'a>> for Message<'a> {
+impl<'a> TryFrom<BorrowedPlainMessageInbound<'a>> for Message<'a> {
     type Error = Error;
 
-    fn try_from(plain: BorrowedPlainMessage<'a>) -> Result<Self, Self::Error> {
+    fn try_from(plain: BorrowedPlainMessageInbound<'a>) -> Result<Self, Self::Error> {
         Ok(Self {
             version: plain.version,
             payload: MessagePayload::new(plain.typ, plain.version, plain.payload)?,
         })
     }
 }
 
+pub trait BorrowedPlainMessage: Sized {
+    fn typ(&self) -> ContentType;
+
+    fn version(&self) -> ProtocolVersion;
+
+    fn payload_to_vec(&self) -> Vec<u8>;
+
+    fn payload_len(&self) -> usize;
+
+    fn encoded_len(&self, record_layer: &RecordLayer) -> usize {
+        OpaqueMessage::HEADER_SIZE as usize + record_layer.encrypted_len(self.payload_len())
+    }
+
+    fn to_unencrypted_opaque(&self) -> OpaqueMessage {
+        OpaqueMessage {
+            version: self.version(),
+            typ: self.typ(),
+            payload: Payload::Owned(self.payload_to_vec()),
+        }
+    }
+
+    fn into_owned(self) -> PlainMessage {
+        PlainMessage {
+            version: self.version(),
+            typ: self.typ(),
+            payload: Payload::Owned(self.payload_to_vec()),
+        }
+    }
+}
+
 /// A TLS frame, named TLSPlaintext in the standard.
 ///
-/// This type differs from `OpaqueMessage` because it borrows
-/// its payload.  You can make a `OpaqueMessage` from an
-/// `BorrowMessage`, but this involves a copy.
+/// This type borrows its decrypted payload from a `MessageDeframer`.
+/// You can make a `OpaqueMessage` from an `BorrowedPlainMessageInbound`,
+/// but this involves a copy.
 ///
 /// This type also cannot decode its internals and
 /// cannot be read/encoded; only `OpaqueMessage` can do that.
 #[derive(Debug)]
-pub struct BorrowedPlainMessage<'a> {
+pub struct BorrowedPlainMessageInbound<'a> {
```

Comment:
Sorry, I didn't see this discussion before posting https://github.com/rustls/rustls/pull/1774#issuecomment-1932096207.

I agree with ctz about `Borrowed` not telling much and I would argue the ownership of a type in general is not helpful. To me it feels like the ownership and mutability (and all other requirements) can be derived from the data format and "direction", but not the other way around.
If we have both `OpaqueMessage` and `OwnedOpaqueMessage` it's not clear why we don't only use the former.
But if we have `InboundOpaqueMessage` and `OutboundOpaqueMessage` it's immediately clear when to use one or the other. The fact that `OutboundOpaqueMessage` owns its data is an implementation detail, we do not use it because it owns its data, we use it to process outgoing messages and its data is owned because the encrypters allocate and copy plaintext before encrypting it.